File: chazm-model/src/main/java/io/github/runtimemodels/chazm/Relations.java
package io.github.runtimemodels.chazm;

import runtimemodels.chazm.api.entity.*;
import runtimemodels.chazm.api.id.UniqueId;
import runtimemodels.chazm.api.relation.*;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

class Relations {

    final Map<UniqueId<Role>, Map<UniqueId<SpecificationGoal>, Achieves>> achieves = new ConcurrentHashMap<>();
    final Map<UniqueId<SpecificationGoal>, Map<UniqueId<Role>, Achieves>> achievedBy = new ConcurrentHashMap<>();
    final Map<UniqueId<Assignment>, Assignment> assignments = new ConcurrentHashMap<>();
    final Map<UniqueId<Agent>, Map<UniqueId<Assignment>, Assignment>> assignmentsByAgent = new ConcurrentHashMap<>();
    final Map<UniqueId<Role>, Map<UniqueId<Characteristic>, Contains>> contains = new ConcurrentHashMap<>();
    final Map<UniqueId<Characteristic>, Map<UniqueId<Role>, Contains>> containedBy = new ConcurrentHashMap<>();
    final Map<UniqueId<Agent>, Map<UniqueId<Attribute>, Has>> has = new ConcurrentHashMap<>();
    final Map<UniqueId<Attribute>, Map<UniqueId<Agent>, Has>> hadBy = new ConcurrentHashMap<>();
    final Map<UniqueId<Pmf>, Moderates> moderates = new ConcurrentHashMap<>();
    final Map<UniqueId<Attribute>, Map<UniqueId<Pmf>, Moderates>> moderatedBy = new ConcurrentHashMap<>();
    final Map<UniqueId<Role>, Map<UniqueId<Attribute>, Needs>> needs = new ConcurrentHashMap<>();
    final Map<UniqueId<Attribute>, Map<UniqueId<Role>, Needs>> neededBy = new ConcurrentHashMap<>();
    final Map<UniqueId<Agent>, Map<UniqueId<Capability>, Possesses>> possesses = new ConcurrentHashMap<>();
    final Map<UniqueId<Capability>, Map<UniqueId<Agent>, Possesses>> possessedBy = new ConcurrentHashMap<>();
    final Map<UniqueId<Role>, Map<UniqueId<Capability>, Requires>> requires = new ConcurrentHashMap<>();
    final Map<UniqueId<Capability>, Map<UniqueId<Role>, Requires>> requiredBy = new ConcurrentHashMap<>();
    final Map<UniqueId<Role>, Map<UniqueId<Pmf>, Uses>> uses = new ConcurrentHashMap<>();
    final Map<UniqueId<Pmf>, Map<UniqueId<Role>, Uses>> usedBy = new ConcurrentHashMap<>();

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/Relations.kt
package runtimemodels.chazm.model

import runtimemodels.chazm.api.id.*
import runtimemodels.chazm.api.relation.Assignment
import runtimemodels.chazm.api.relation.Requires
import runtimemodels.chazm.api.relation.Uses
import java.util.concurrent.ConcurrentHashMap

internal class Relations {

    //    val achieves: MutableMap<RoleId, MutableMap<SpecificationGoalId, Achieves>> = ConcurrentHashMap()
//    val achievedBy: MutableMap<SpecificationGoalId, MutableMap<RoleId, Achieves>> = ConcurrentHashMap()
    val assignments: MutableMap<UniqueId<Assignment>, Assignment> = ConcurrentHashMap()
    val assignmentsByAgent: MutableMap<AgentId, MutableMap<UniqueId<Assignment>, Assignment>> = ConcurrentHashMap()
    //    val contains: MutableMap<RoleId, MutableMap<CharacteristicId, Contains>> = ConcurrentHashMap()
//    val containedBy: MutableMap<CharacteristicId, MutableMap<RoleId, Contains>> = ConcurrentHashMap()
//    val has: MutableMap<AgentId, MutableMap<AttributeId, Has>> = ConcurrentHashMap()
//    val hadBy: MutableMap<AttributeId, MutableMap<AgentId, Has>> = ConcurrentHashMap()
//    val moderates: MutableMap<PmfId, Moderates> = ConcurrentHashMap()
//    val moderatedBy: MutableMap<AttributeId, MutableMap<PmfId, Moderates>> = ConcurrentHashMap()
//    val needs: MutableMap<RoleId, MutableMap<AttributeId, Needs>> = ConcurrentHashMap()
//    val neededBy: MutableMap<AttributeId, MutableMap<RoleId, Needs>> = ConcurrentHashMap()
//    val possesses: MutableMap<AgentId, MutableMap<CapabilityId, Possesses>> = ConcurrentHashMap()
//    val possessedBy: MutableMap<CapabilityId, MutableMap<AgentId, Possesses>> = ConcurrentHashMap()
    val requires: MutableMap<RoleId, MutableMap<CapabilityId, Requires>> = ConcurrentHashMap()
    val requiredBy: MutableMap<CapabilityId, MutableMap<RoleId, Requires>> = ConcurrentHashMap()
    val uses: MutableMap<RoleId, MutableMap<PmfId, Uses>> = ConcurrentHashMap()
    val usedBy: MutableMap<PmfId, MutableMap<RoleId, Uses>> = ConcurrentHashMap()

}
